### PR TITLE
[#346] advanced slot settings applying

### PR DESCRIPTION
**Pull request by Pavliuk13**
https://user-images.githubusercontent.com/59260347/190870832-decb8c93-1a69-422d-a158-25a5c253989d.mp4

### CODE REVIEW

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 1]
```diff
@@ -38,31 +41,46 @@ export class GeneralComponent extends BaseComponent implements OnInit {
         };
         this.advancedSettings = {
             maxNumberOfBookings: this.slot?.advancedSlotSettings?.maxNumberOfBookings ?? 1,
-            frequency: this.slot?.advancedSlotSettings?.frequency ?? this.slotsFrequencies[0],
+            frequency: this.slot?.advancedSlotSettings?.frequency ?? this.durations[0].minutes!,
             days: this.slot?.advancedSlotSettings?.days ?? 1,
-            paddingMeeting: this.slot?.advancedSlotSettings?.paddingMeeting ?? this.meetingPaddings[0],
+            paddingMeeting: this.slot?.advancedSlotSettings?.paddingMeeting ?? this.durations[0].minutes!,
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: Could you avoid !

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 2]
```diff
@@ -38,31 +41,46 @@ export class GeneralComponent extends BaseComponent implements OnInit {
...
             activityType: this.slot?.advancedSlotSettings?.activityType ?? this.slotActivityOptionsEnums[0],
             minBookingMeetingDifference:
-                this.slot?.advancedSlotSettings?.minBookingMeetingDifference ?? this.minBookingMeetingDifferences[0],
+                this.slot?.advancedSlotSettings?.minBookingMeetingDifference ?? this.durations[0].minutes!,
+            startDate: this.slot?.advancedSlotSettings?.startDate ?? new Date(new Date().setHours(0, 0, 0, 0)),
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: Please our date library for this (startOfDay)

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 3]
```diff
@@ -116,13 +128,16 @@ export class GeneralComponent extends BaseComponent implements OnInit {
         this.advancedSettings = {
             maxNumberOfBookings: 1,
             days: 1,
-            paddingMeeting: this.meetingPaddings[0],
+            paddingMeeting: this.durations[0].minutes!,
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: see the previous comment

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 4]
```diff
@@ -116,13 +128,16 @@ export class GeneralComponent extends BaseComponent implements OnInit {
...
             activityType: this.slotActivityOptionsEnums[0],
-            minBookingMeetingDifference: this.minBookingMeetingDifferences[0],
-            frequency: this.slotsFrequencies[0],
+            minBookingMeetingDifference: this.durations[0].minutes!,
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: see the previous comment

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 5]
```diff
@@ -135,45 +150,107 @@ export class GeneralComponent extends BaseComponent implements OnInit {
         this.settings.name = removeExcessiveSpaces(value);
     }
 
-    onDurationChange() {
-        this.customTimeShown = this.duration.time === 'Custom';
-        this.settings = this.updateSettingSize(this.duration.minutes!);
+    onValueChange(index: number) {
+        this.inputSettings[index].isCustom = this.inputSettings[index].durationValue?.time === 'Custom';
+        this.updateSettings(this.inputSettings[index].durationValue?.minutes!, this.inputSettings[index].inputType);
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: Could we get rid of the index? and why do we need it?

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 6]
```diff
@@ -135,45 +150,107 @@ export class GeneralComponent extends BaseComponent implements OnInit {
...
+    }
+
+    customBookingDifferenceChanged(value: number) {
+        this.updateSettings(value, InputFieldType.MinBookTime);
+    }
+
+    saveRange() {
+        this.advancedSettings = {
+            ...this.advancedSettings,
+            startDate: new Date(this.startDate.setHours(0, 0, 0, 0)),
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: see the previous comments

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 7]
```diff
@@ -135,45 +150,107 @@ export class GeneralComponent extends BaseComponent implements OnInit {
...
 
-    private defineCurrentDuration(slotDuration: number) {
-        this.duration =
-            this.durations.find((x) => x.minutes === slotDuration) ?? this.durations.find((x) => x.time === 'Custom')!;
+    private defineInputKeys() {
+        this.durationKey = this.inputSettings.indexOf(this.inputSettings.find((el) => el.inputType === InputFieldType.Duration)!);
+        this.frequencyKey = this.inputSettings.indexOf(this.inputSettings.find((el) => el.inputType === InputFieldType.Frequency)!);
+        this.paddingKey = this.inputSettings.indexOf(this.inputSettings.find((el) => el.inputType === InputFieldType.Padding)!);
+        this.minBookingKey = this.inputSettings.indexOf(this.inputSettings.find((el) => el.inputType === InputFieldType.MinBookTime)!);
+    }
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: looks strange

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 8]
```diff
@@ -135,45 +150,107 @@ export class GeneralComponent extends BaseComponent implements OnInit {
...
+        this.inputSettings = [
+            ...this.inputSettings.map((val) => {
+                val.durationValue =
+                    this.durations.find((x) => x.minutes === this.inputValues[val.inputType]) ??
+                    this.durations.find((x) => x.time === 'Custom')!;
+
+                if (val.durationValue?.time === 'Custom') {
+                    val.isCustom = true;
+                    val.inputValue = String(this.inputValues[val.inputType]);
+                }
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: Please refactor

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 9]
```diff
@@ -97,12 +112,23 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
+                this.padding = this.slot?.advancedSlotSettings?.paddingMeeting ?? 0;
                 this.getOrderedTimes(this.slot!.id);
                 this.selectedMeetingDuration = this.slot!.size;
+                this.frequency = this.slot?.advancedSlotSettings?.frequency ?? this.selectedMeetingDuration;
+                this.minBookingDifference = this.slot?.advancedSlotSettings?.minBookingMeetingDifference ?? 0;
                 this.scheduleItems = changeScheduleItemsDate(resp!.schedule!.scheduleItems);
-                this.disabledDays = resp!
-                    .schedule!.scheduleItems.filter((el) => !el.isEnabled)
-                    .map((el) => WeekDay[el.weekDay]);
+                this.disabledDays = this.scheduleItems.filter((el) => !el.isEnabled).map((el) => WeekDay[el.weekDay]);
```
Comment (tatianahutii) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: Toooo long method

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 10]
```diff
@@ -123,6 +149,16 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
             .pipe(this.untilThis)
             .subscribe((result) => {
                 this.orderedTimes = result;
+
+                const maxBookingsCount = this.slot?.advancedSlotSettings?.maxNumberOfBookings ?? 0;
+
+                this.isBookingsLimit = maxBookingsCount !== 0 && this.orderedTimes.length >= maxBookingsCount;
```
Comment (tatianahutii) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]:
```suggestion
                this.isBookingsLimit = !!maxBookingsCount && this.orderedTimes.length >= maxBookingsCount;
```

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 11]
```diff
@@ -123,6 +149,16 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
+
+                if (this.isBookingsLimit) {
+                    this.notificationService.showInfoMessage(
+                        "The limit on the number of bookings has been reached. You can't book meeting now.",
+                    );
+                }
```
Comment (tatianahutii) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: Move to a separate method

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 12]
```diff
@@ -238,10 +282,20 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
         max.setMonth(result.getMonth());
 
+        if (this.activityType !== ActivityType.Indefinitely) {
+            return (
+                result.getTime() >= min.getTime() &&
+                result.getTime() <= max.getTime() &&
+                result.getTime() >= this.startDate?.getTime()! &&
+                result.getTime() <= this.finishDate?.getTime()! &&
+                differenceInMinutes(result, this.nowDate) > this.minBookingDifference
+            );
```
Comment (tatianahutii) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: please get rid of !, refactor and move to a separate method

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/availability/new-slot/general/custom-time/custom-time.component.sass [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+    p
+        font-weight: bold
+        margin-bottom: 5px
+
+    .custom-time
+        display: flex
+        align-items: flex-start
+        justify-content: space-between
+        max-width: calc(280px - 9px)
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/custom-time/custom-time.component.sass]: I think we can simply write 271px)

File: frontend/src/app/modules/availability/new-slot/general/custom-time/custom-time.component.sass [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        width: calc(100% - 9px)
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/custom-time/custom-time.component.sass]: Why we need this?

File: frontend/src/app/modules/availability/new-slot/general/general.component.html [hunk 15]
```diff
@@ -19,41 +19,20 @@
...
-                        <mat-option *ngFor="let unit of unitOfTime" [value]="unit">
-                            {{ unit }}
-                        </mat-option>
-                    </mat-select>
-                </div>
-                <div *ngIf="customTime.invalid && customTime.touched">
-                    <small *ngIf="customTimeShown && customTime.errors?.['required']">Field is required</small>
-                    <small *ngIf="customTime.errors?.['pattern']">Input field accepts only numbers</small>
-                </div>
+            <!-- custom duration for duration -->
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/general.component.html]: Do we need this comment?

File: frontend/src/app/modules/availability/new-slot/general/general.component.html [hunk 16]
```diff
@@ -19,41 +19,20 @@
...
+            <div class="input-container" *ngIf="inputSettings[durationKey].isCustom">
+                <app-custom-time
+                    [inputCustomTime]="inputSettings[durationKey].inputValue"
+                    (changeValueEmitter)="customDurationChanged($event)"></app-custom-time>
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/general.component.html]: Closing tag should be on the next line) It doesn't look beatiful)

File: frontend/src/app/modules/availability/new-slot/general/general.component.html [hunk 17]
```diff
@@ -119,43 +98,79 @@
...
+                    [(ngModel)]="inputSettings[frequencyKey].durationValue"
+                    [disabled]="!addAdvanced"
+                    name="slotFrequency"
+                    (selectionChange)="onValueChange(frequencyKey)">
+                    <mat-option *ngFor="let _duration of durations" [value]="_duration">
+                        {{ _duration.time }} {{ _duration.unitOfTime }}
                     </mat-option>
                 </mat-select>
             </div>
+            <!-- custom duration for frequency -->
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/general.component.html]: Same

File: frontend/src/app/modules/availability/new-slot/general/general.component.html [hunk 18]
```diff
@@ -119,43 +98,79 @@
...
+            <div class="input-container" *ngIf="inputSettings[frequencyKey].isCustom">
+                <app-custom-time
+                    [inputCustomTime]="inputSettings[frequencyKey].inputValue"
+                    (changeValueEmitter)="customFrequencyChanged($event)"></app-custom-time>
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/general.component.html]: Same

File: frontend/src/app/modules/availability/new-slot/general/general.component.html [hunk 19]
```diff
@@ -119,43 +98,79 @@
...
-                    [disabled]="!addAdvanced">
-                    <mat-option *ngFor="let meetingPadding of meetingPaddings" [value]="meetingPadding">
-                        {{ meetingPadding }} min
+                    (selectionChange)="onValueChange(paddingKey)">
+                    <mat-option *ngFor="let _duration of durations" [value]="_duration">
+                        {{ _duration.time }} {{ _duration.unitOfTime }}
                     </mat-option>
                 </mat-select>
             </div>
+            <!-- custom duration for paddings -->
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/general.component.html]: Same

File: frontend/src/app/modules/availability/new-slot/general/general.component.html [hunk 20]
```diff
@@ -119,43 +98,79 @@
...
+            <div class="input-container" *ngIf="inputSettings[paddingKey].isCustom">
+                <app-custom-time
+                    [inputCustomTime]="inputSettings[paddingKey].inputValue"
+                    (changeValueEmitter)="customPaddingChanged($event)"></app-custom-time>
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/general.component.html]: Same

File: frontend/src/app/modules/availability/new-slot/general/general.component.sass [hunk 21]
```diff
@@ -58,7 +58,7 @@
     @media ( max-width: $desktop-min-width)
         margin-top: 0
 
-input[type=text], .mat-select, .label-container, .error
+input[type=text], .mat-select, .label-container, .error, .range-wrapper
     max-width: calc(280px - 9px)
```
Comment (Erbidium) [inline frontend/src/app/modules/availability/new-slot/general/general.component.sass]: Simply 271px

**Review by Erbidium — CHANGES_REQUESTED** [review]

**Review by Erbidium — APPROVED** [review]

File: frontend/src/app/modules/availability/new-slot/general/custom-time/custom-time.component.sass [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    font-size: 14px
+    padding-left: 5px
+    border: 2px solid $pattens-blue
+    border-radius: 8px
+    height: 36px
+
+.ng-invalid.ng-touched:not(.ng-pristine)
+    border-color: $red
+
+::ng-deep .input-container
```
Comment (apofis96) [inline frontend/src/app/modules/availability/new-slot/general/custom-time/custom-time.component.sass]: input-container are likely to be used in other component

**Review by apofis96 — CHANGES_REQUESTED** [review]

**Review by apofis96 — APPROVED** [review]

**Review by tatianahutii — APPROVED** [review]

**Pull request merged**